### DISCUSSION TITLE
Sharing WebApplicationFactory between test classes

### DISCUSSION THREAD

**Discussion by tvytrykush** [xunit/xunit#2893] [Question]
Hello XUnit community, I'm writing now integration tests and after adding second test class I've noticed that the lifecycle behavior isn't what I would expect it to be.<br> I use <b>Testcontainers</b> to spawn containerized dependencies, <b>WebApplicationFactory</b> as service provider and for all the heavy lifting and <b>XUnit</b> `IAsyncLifecycle` to control the lifespan of it. <br>
I've got CustomWebApplicationFactory with implemented `IAsyncLifecycle`:
```csharp
 public class CustomWebApplicationFactory<TStartup>
   (SharedKeycloakContainer keycloakContainer,
    SharedSqlContainer sqlContainer, 
    // ... other containers
    )
   : WebApplicationFactory<TStartup>, IAsyncLifetime where TStartup: class
{
   protected override void ConfigureWebhost(IWebHostBuilder builder)
   {
       // Irrelevant code
   }

   public Task InitializeAsync() 
      => Task.WhenAll(keycloakContainer.Instance.StartAsync(), sqlContainer.Instance.StartAsync()) 

   public Task DisposeAsync() 
      => Task.WhenAll(keycloakContainer.Instance.StopAsync(), sqlContainer.Instance.StopAsync()) 
}
```

Also I have a collection definition:
```csharp
[CollectionDefinition(nameof(SharedTestInfrastructureCollection))]
public class SharedTestInfrastructureCollection 
    : ICollectionFixture<SharedKeycloakContainer>,
        ICollectionFixture<SharedSqlContainer>
{
}
```

My containers are just wrappers with some custom actions required to do some setup, but for major part they're just like this:
```csharp
public class SharedKeycloakContainer
{
   public KeycloakContainer Instance { get; } = new KeycloakBuilder()
        .WithName("keycloak-integration-test")
        .Build();
}
```

And an integration test base class:
```csharp
[Collection(nameof(SharedTestInfrastructureCollection))]
public class IntegrationTestBase(CustomWebApplicationFactory<IApiMarker> factory)
    : IClassFixture<CustomWebApplicationFactory<IApiMarker>>
{
    protected Faker FakerCommon = new();
    protected ExtendedHttpClient HttpClient { get; } = new(factory.CreateClient());
    protected CustomWebApplicationFactory<IApiMarker> Factory { get; } = factory;
    protected CustomDbContext DbContext { get; } = factory.Services.GetRequiredService<CustomDbContext>();
}
```

And finally my test class:
```csharp
public class DevicesControllerTests(
    CustomWebApplicationFactory<IApiMarker> factory)
    : IntegrationTestBase(factory)
{
    [Fact]
    public async Task ShouldDoSomethingUseful()
    {
        using var scope = Factory.Services.CreateScope();
        // ...............
        // Irrelevant code
    }
}
```

Now what I would expect with such setup is that it should create single CustomWebApplicationFactory for all tests that inherit from IntegrationTestBase and reuse containers and IHost. What I get is that every test class spawns new containers and new WebApplicationFactory what is a complete waste of resources, given that I need to wait for ~2 minutes to have all required containers ready and really would like to have just a single instance of this whole setup for arbitrary amount of test classes. 
What I've tried and failed:
1. Apply `[Collection(nameof(SharedTestInfrastructureCollection))]` attribute to the test class itself - obviously with no difference in behavior.
2. Try create separate fixture that will create CustomWebApplicationFactory from constuctor and also pass containers as arguments(though it's hard then to correctly wait till containers are started) - I can't properly control containers startup with this approach.

Overall, maybe I move in completely wrong direction and should use something else instead to achieve the required behavior mentioned above. Would be very greatful for any suggestions.

**Comment by bradwilson** [comment]
> Now what I would expect with such setup is that it should create single CustomWebApplicationFactory for all tests that inherit from IntegrationTestBase and reuse containers and IHost.

Unfortunately this expectation is incorrect. What you're describing is a collection fixture, where all the test classes in question share a collection. The fact that `IClassFixture` is decorated on the base class does not cause them to be shared; instead, you've created several classes from the base type which all get their own copies of `CustomWebApplicationFactory<IApiMarker>`.

In essence, these two sets of declarations are the same w/rt the lifetime & ownership of the class fixture instances:

```csharp
public class BaseClass : IClassFixture<Foo> { }
public class TestClass1 : BaseClass { } 
public class TestClass2 : BaseClass { }
```

```csharp
public class TestClass1 : IClassFixture<Foo> { }
public class TestClass2 : IClassFixture<Foo> { }
```

In both cases, `TestClass1` and `TestClass2` will get their own unique instances of `Foo` (all the test methods in `TestClass1` will share one instance, and all the test methods in `TestClass2` will share a different instance).

You should add `ICollectionFixture<CustomWebApplicationFactory<IApiMarker>>` to `SharedTestInfrastructureCollection` instead.

Documentation for shared context: https://xunit.net/docs/shared-context

----------

It's worth pointing out, if you're thinking about trying to move the inheritance up to the test collection definition class, that _all the same rules apply_; that is, these two things are identical:

```csharp
public class BaseCollection : ICollectionFixture<Foo> { }

[CollectionDefinition(nameof(Collection1))]
public class Collection1 : BaseCollection { }

[CollectionDefinition(nameo(Collection2))]
public class Collection2 : BaseCollection { }
```

```csharp
[CollectionDefinition(nameof(Collection1))]
public class Collection1 : ICollectionFixture<Foo> { }

[CollectionDefinition(nameo(Collection2))]
public class Collection2 : ICollectionFixture<Foo> { }
```

In both cases, `Collection1` and `Collection2` will get unique instances of `Foo`. Test methods in test classes in `Collection1` will share one instance, while test methods in test classes in `Collection2` will share a difference instance.

Additionally, there is no feature for "hierarchy" in fixtures at the same level; that is, there's no such thing as a class fixture that's shared across multiple classes and/or as a dependency from another class fixture. The only way to "share" something is to move it up a layer: class fixtures can share an instance of a collection fixture (and in v3, collection fixtures can share an instance of an assembly fixture).
